File: gateway/internal/store/providers.go
package store

import (
	"context"
	"fmt"
	"time"
)

type LLMProvider struct {
	ID          string    `json:"id"`
	Key         string    `json:"key"`
	DisplayName string    `json:"display_name"`
	APIType     string    `json:"api_type"`
	BaseURL     string    `json:"base_url"`
	APIKeyRef   string    `json:"api_key_ref"`
	Enabled     bool      `json:"enabled"`
	SortOrder   int       `json:"sort_order"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type LLMModel struct {
	ID            string    `json:"id"`
	ProviderKey   string    `json:"provider_key"`
	ModelID       string    `json:"model_id"`
	DisplayName   string    `json:"display_name"`
	ContextWindow int       `json:"context_window"`
	MaxTokens     int       `json:"max_tokens"`
	Reasoning     bool      `json:"reasoning"`
	Enabled       bool      `json:"enabled"`
	SortOrder     int       `json:"sort_order"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

func (s *Store) ListProviders(ctx context.Context) ([]LLMProvider, error) {
	rows, err := s.pool.Query(ctx,
		`SELECT id, key, display_name, api_type, base_url, api_key_ref, enabled, sort_order, created_at, updated_at
		 FROM llm_providers ORDER BY sort_order, key`)
	if err != nil {
		return nil, fmt.Errorf("list providers: %w", err)
	}
	defer rows.Close()

	var providers []LLMProvider
	for rows.Next() {
		var p LLMProvider
		if err := rows.Scan(&p.ID, &p.Key, &p.DisplayName, &p.APIType, &p.BaseURL, &p.APIKeyRef, &p.Enabled, &p.SortOrder, &p.CreatedAt, &p.UpdatedAt); err != nil {
			return nil, fmt.Errorf("scan provider: %w", err)
		}
		providers = append(providers, p)
	}
	return providers, nil
}

func (s *Store) GetProvider(ctx context.Context, key string) (*LLMProvider, error) {
	p := &LLMProvider{}
	err := s.pool.QueryRow(ctx,
		`SELECT id, key, display_name, api_type, base_url, api_key_ref, enabled, sort_order, created_at, updated_at
		 FROM llm_providers WHERE key = $1`, key,
	).Scan(&p.ID, &p.Key, &p.DisplayName, &p.APIType, &p.BaseURL, &p.APIKeyRef, &p.Enabled, &p.SortOrder, &p.CreatedAt, &p.UpdatedAt)
	if err != nil {
		return nil, fmt.Errorf("get provider %q: %w", key, err)
	}
	return p, nil
}

func (s *Store) UpsertProvider(ctx context.Context, p *LLMProvider) error {
	_, err := s.pool.Exec(ctx,
		`INSERT INTO llm_providers (key, display_name, api_type, base_url, api_key_ref, enabled, sort_order)
		 VALUES ($1, $2, $3, $4, $5, $6, $7)
		 ON CONFLICT (key) DO UPDATE SET
		   display_name = EXCLUDED.display_name,
		   api_type = EXCLUDED.api_type,
		   base_url = EXCLUDED.base_url,
		   api_key_ref = EXCLUDED.api_key_ref,
		   enabled = EXCLUDED.enabled,
		   sort_order = EXCLUDED.sort_order`,
		p.Key, p.DisplayName, p.APIType, p.BaseURL, p.APIKeyRef, p.Enabled, p.SortOrder,
	)
	if err != nil {
		return fmt.Errorf("upsert provider %q: %w", p.Key, err)
	}
	return nil
}

func (s *Store) DeleteProvider(ctx context.Context, key string) error {
	tag, err := s.pool.Exec(ctx,
		`DELETE FROM llm_providers WHERE key = $1`, key)
	if err != nil {
		return fmt.Errorf("delete provider %q: %w", key, err)
	}
	if tag.RowsAffected() == 0 {
		return fmt.Errorf("provider not found: %s", key)
	}
	return nil
}

func (s *Store) ListModels(ctx context.Context) ([]LLMModel, error) {
	rows, err := s.pool.Query(ctx,
		`SELECT id, provider_key, model_id, display_name, context_window, max_tokens, reasoning, enabled, sort_order, created_at, updated_at
		 FROM llm_models ORDER BY provider_key, sort_order, model_id`)
	if err != nil {
		return nil, fmt.Errorf("list models: %w", err)
	}
	defer rows.Close()

	var models []LLMModel
	for rows.Next() {
		var m LLMModel
		if err := rows.Scan(&m.ID, &m.ProviderKey, &m.ModelID, &m.DisplayName, &m.ContextWindow, &m.MaxTokens, &m.Reasoning, &m.Enabled, &m.SortOrder, &m.CreatedAt, &m.UpdatedAt); err != nil {
			return nil, fmt.Errorf("scan model: %w", err)
		}
		models = append(models, m)
	}
	return models, nil
}

func (s *Store) ListModelsByProvider(ctx context.Context, providerKey string) ([]LLMModel, error) {
	rows, err := s.pool.Query(ctx,
		`SELECT id, provider_key, model_id, display_name, context_window, max_tokens, reasoning, enabled, sort_order, created_at, updated_at
		 FROM llm_models WHERE provider_key = $1 ORDER BY sort_order, model_id`,
		providerKey,
	)
	if err != nil {
		return nil, fmt.Errorf("list models by provider %q: %w", providerKey, err)
	}
	defer rows.Close()

	var models []LLMModel
	for rows.Next() {
		var m LLMModel
		if err := rows.Scan(&m.ID, &m.ProviderKey, &m.ModelID, &m.DisplayName, &m.ContextWindow, &m.MaxTokens, &m.Reasoning, &m.Enabled, &m.SortOrder, &m.CreatedAt, &m.UpdatedAt); err != nil {
			return nil, fmt.Errorf("scan model: %w", err)
		}
		models = append(models, m)
	}
	return models, nil
}

func (s *Store) UpsertModel(ctx context.Context, m *LLMModel) error {
	_, err := s.pool.Exec(ctx,
		`INSERT INTO llm_models (provider_key, model_id, display_name, context_window, max_tokens, reasoning, enabled, sort_order)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
		 ON CONFLICT (provider_key, model_id) DO UPDATE SET
		   display_name = EXCLUDED.display_name,
		   context_window = EXCLUDED.context_window,
		   max_tokens = EXCLUDED.max_tokens,
		   reasoning = EXCLUDED.reasoning,
		   enabled = EXCLUDED.enabled,
		   sort_order = EXCLUDED.sort_order`,
		m.ProviderKey, m.ModelID, m.DisplayName, m.ContextWindow, m.MaxTokens, m.Reasoning, m.Enabled, m.SortOrder,
	)
	if err != nil {
		return fmt.Errorf("upsert model %q/%q: %w", m.ProviderKey, m.ModelID, err)
	}
	return nil
}

func (s *Store) DeleteModel(ctx context.Context, id string) error {
	tag, err := s.pool.Exec(ctx,
		`DELETE FROM llm_models WHERE id = $1`, id)
	if err != nil {
		return fmt.Errorf("delete model %q: %w", id, err)
	}
	if tag.RowsAffected() == 0 {
		return fmt.Errorf("model not found: %s", id)
	}
	return nil
}

File: website/src/routes/admin/providers.rs
#![allow(dead_code)]
// Legacy standalone provider page code is retained while GET /admin/providers redirects into the consolidated config center.

use askama::Template;
use axum::extract::{Path, State};
use axum::http::HeaderMap;
use axum::response::{IntoResponse, Redirect, Response};

use crate::auth::AdminUser;
use crate::error::{render_template, AppError};
use crate::models::provider;
use crate::AppState;

// ─── Views ───────────────────────────────────────────

#[allow(dead_code)]
struct ProviderView {
    key: String,
    display_name: String,
    api_type: String,
    base_url: String,
    api_key_ref: String,
    enabled: bool,
    sort_order: i32,
}

#[allow(dead_code)]
struct ModelView {
    id: String,
    provider_key: String,
    model_id: String,
    display_name: String,
    context_window: i32,
    max_tokens: i32,
    reasoning: bool,
    enabled: bool,
    sort_order: i32,
}

// ─── Templates ───────────────────────────────────────

#[derive(Template)]
#[template(path = "admin/providers.html")]
struct ProvidersTemplate {
    admin_name: String,
    active_page: String,
    providers: Vec<ProviderView>,
    models: Vec<ModelView>,
    save_success: bool,
    save_error: Option<String>,
}

// ─── List ────────────────────────────────────────────

#[derive(serde::Deserialize)]
pub struct ListQuery {
    success: Option<bool>,
    error: Option<String>,
}

pub async fn list(
    State(state): State<AppState>,
    admin: AdminUser,
    axum::extract::Query(query): axum::extract::Query<ListQuery>,
) -> Result<impl IntoResponse, AppError> {
    render_providers_page(&state, &admin, query.success.unwrap_or(false), query.error).await
}

// ─── Upsert Provider ────────────────────────────────

#[derive(serde::Deserialize)]
pub struct ProviderForm {
    key: String,
    display_name: String,
    api_type: String,
    base_url: String,
    api_key_ref: String,
    enabled: Option<String>,
}

pub async fn upsert_provider_handler(
    State(state): State<AppState>,
    _admin: AdminUser,
    axum::Form(form): axum::Form<ProviderForm>,
) -> Result<Response, AppError> {
    let key = form.key.trim();
    if key.is_empty() {
        return Ok(Redirect::to("/admin/providers?error=Provider key 不能为空").into_response());
    }

    let enabled = form.enabled.is_some();
    provider::upsert_provider(
        &state.db,
        key,
        form.display_name.trim(),
        form.api_type.trim(),
        form.base_url.trim(),
        form.api_key_ref.trim(),
        enabled,
    )
    .await?;

    Ok(Redirect::to("/admin/providers?success=true").into_response())
}

// ─── Delete Provider ────────────────────────────────

pub async fn delete_provider_handler(
    State(state): State<AppState>,
    _admin: AdminUser,
    headers: HeaderMap,
    Path(key): Path<String>,
) -> Result<Response, AppError> {
    provider::delete_provider(&state.db, &key).await?;

    if is_htmx_request(&headers) {
        Ok(axum::response::Html("").into_response())
    } else {
        Ok(Redirect::to("/admin/providers?success=true").into_response())
    }
}

// ─── Upsert Model ───────────────────────────────────

#[derive(serde::Deserialize)]
pub struct ModelForm {
    provider_key: String,
    model_id: String,
    display_name: String,
    context_window: i32,
    max_tokens: i32,
    reasoning: Option<String>,
    enabled: Option<String>,
}

pub async fn upsert_model_handler(
    State(state): State<AppState>,
    _admin: AdminUser,
    axum::Form(form): axum::Form<ModelForm>,
) -> Result<Response, AppError> {
    let provider_key = form.provider_key.trim();
    let model_id = form.model_id.trim();
    if provider_key.is_empty() || model_id.is_empty() {
        return Ok(
            Redirect::to("/admin/providers?error=Provider key 和 Model ID 不能为空")
                .into_response(),
        );
    }

    let reasoning = form.reasoning.is_some();
    let enabled = form.enabled.is_some();
    provider::upsert_model(
        &state.db,
        provider_key,
        model_id,
        form.display_name.trim(),
        form.context_window,
        form.max_tokens,
        reasoning,
        enabled,
    )
    .await?;

    Ok(Redirect::to("/admin/providers?success=true").into_response())
}

// ─── Delete Model ───────────────────────────────────

pub async fn delete_model_handler(
    State(state): State<AppState>,
    _admin: AdminUser,
    headers: HeaderMap,
    Path(id): Path<String>,
) -> Result<Response, AppError> {
    provider::delete_model(&state.db, &id).await?;

    if is_htmx_request(&headers) {
        Ok(axum::response::Html("").into_response())
    } else {
        Ok(Redirect::to("/admin/providers?success=true").into_response())
    }
}

// ─── Helpers ─────────────────────────────────────────

async fn render_providers_page(
    state: &AppState,
    admin: &AdminUser,
    save_success: bool,
    save_error: Option<String>,
) -> Result<impl IntoResponse, AppError> {
    let db_providers = provider::list_providers(&state.db).await?;
    let db_models = provider::list_models(&state.db).await?;

    let providers: Vec<ProviderView> = db_providers
        .into_iter()
        .map(|p| ProviderView {
            key: p.key,
            display_name: p.display_name,
            api_type: p.api_type,
            base_url: p.base_url,
            api_key_ref: p.api_key_ref,
            enabled: p.enabled,
            sort_order: p.sort_order,
        })
        .collect();

    let models: Vec<ModelView> = db_models
        .into_iter()
        .map(|m| ModelView {
            id: m.id,
            provider_key: m.provider_key,
            model_id: m.model_id,
            display_name: m.display_name,
            context_window: m.context_window,
            max_tokens: m.max_tokens,
            reasoning: m.reasoning,
            enabled: m.enabled,
            sort_order: m.sort_order,
        })
        .collect();

    render_template(&ProvidersTemplate {
        admin_name: admin.0.name.clone(),
        active_page: "providers".into(),
        providers,
        models,
        save_success,
        save_error,
    })
}

fn is_htmx_request(headers: &HeaderMap) -> bool {
    headers
        .get("HX-Request")
        .and_then(|v| v.to_str().ok())
        .map(|v| v.eq_ignore_ascii_case("true"))
        .unwrap_or(false)
}
